### PR TITLE
SABER Level 1 Support in wolfSSH

### PR DESCRIPTION
**Pull request by anhu**
Tested with the following: 

./autogen.sh 
./configure --enable-liboqs
make all check 

Also tested for interop with OQS fork of libssh. See README.md for instructions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ wolfSSL. The following is the simplest configuration of wolfSSL to
 enable wolfSSH.
 
     $ cd wolfssl
-    $ ./configure [OPTIONS] --enable-ssh
+    $ ./configure [OPTIONS] --enable-wolfssh
```
Comment (dgarske) [inline README.md:16]: This option was only added recently in https://github.com/wolfSSL/wolfssl/pull/5100 . For now please leave the `--enable-ssh`.

File: wolfssh/internal.h [hunk 2]
```diff
@@ -467,6 +479,7 @@ typedef struct HandshakeInfo {
 #endif
 
     byte useEcc;
+    byte useSaber;
```
Comment (dgarske) [inline wolfssh/internal.h:482]: Please wrap with `#ifndef WOLFSSH_NO_SABER_LEVEL1_SHA256`. Also these should be bit-fields. `byte useSaber:1;` and `byte useEcc:1;`. You'll also need all the code with `!ssh->handshake->useSaber)` wrapped in the macro. I don't want code size to grow if saber is not used.

File: wolfssh/internal.h [hunk 3]
```diff
@@ -404,6 +415,7 @@ struct WOLFSSH_CTX {
     byte* privateKey;                 /* Owned by CTX */
     word32 privateKeySz;
     byte useEcc;                      /* Depends on the private key */
+    byte useSaber;                    /* Depends on the private key */
```
Comment (dgarske) [inline wolfssh/internal.h:418]: Please wrap with `#ifndef WOLFSSH_NO_SABER_LEVEL1_SHA256`. Also these should be bit-fields. `byte useSaber:1;` and `byte useEcc:1;`. You'll also need all the code with `!ssh->handshake->useSaber)` wrapped in the macro. I don't want code size to grow if saber is not used.

File: src/internal.c [hunk 4]
```diff
@@ -7977,15 +8086,43 @@ int SendKexDhInit(WOLFSSH* ssh)
 #else
             ret = WS_INVALID_ALGO_ID;
 #endif /* !defined(WOLFSSH_NO_ECDH) */
+        } else if (ssh->handshake->useSaber) {
```
Comment (dgarske) [inline src/internal.c:8089]: Please use:

```
}
#ifndef WOLFSSH_NO_SABER_LEVEL1_SHA256
else if (ssh->handshake->useSaber) {
...
}
#endif
```

**Review by dgarske — CHANGES_REQUESTED** [review]
Please wrap all `useSaber` code in macros. We don't want that code present when Saber / Liboqs is not used.

File: wolfssh/internal.h (hunk 2 above)
Comment (ejohnstown) [reply wolfssh/internal.h:482]: Do use a bit-field with useSaber, but don't change useEcc yet. That's off topic for this PR. Doing it right when adding it is good. I don't want to change things piecemeal. I'd rather we convert the flags all at once. (The single bit flags aren't the worst memory wasters in wolfSSH, anyway.)

File: src/internal.c [hunk 5]
```diff
@@ -7977,15 +8103,45 @@ int SendKexDhInit(WOLFSSH* ssh)
 #else
             ret = WS_INVALID_ALGO_ID;
 #endif /* !defined(WOLFSSH_NO_ECDH) */
+#ifndef WOLFSSH_NO_SABER_LEVEL1_SHA256
+        } else if (ssh->handshake->useSaber) {
+            OQS_KEM* kem = NULL;
+            ret = 0;
+
+            if (ret == 0) {
```
Comment (dgarske) [inline src/internal.c:8111]: `ret` is always 0 here...

File: wolfssh/internal.h [hunk 6]
```diff
@@ -404,6 +417,9 @@ struct WOLFSSH_CTX {
     byte* privateKey;                 /* Owned by CTX */
     word32 privateKeySz;
     byte useEcc;                      /* Depends on the private key */
+#ifndef WOLFSSH_NO_SABER_LEVEL1_SHA256
+    byte useSaber:1;                    /* Depends on the private key */
```
Comment (dgarske) [inline wolfssh/internal.h:421]: NIT: Line up comments.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/internal.c (hunk 5 above)
Comment (anhu) [reply src/internal.c:8111]: But it looks so nice and consistent.  Anyways, fixed.  Stay tuned.

**Review by dgarske — APPROVED** [review]
Looks good! Thanks. Over to John for final.

File: wolfssh/internal.h (hunk 2 above)
Comment (anhu) [reply wolfssh/internal.h]: OH...thats an interesting point.  If this gets expanded to different levels (ie 1, 3, 5) then I can switch this back to BYTE.

**Review by ejohnstown — APPROVED** [review]

**Pull request merged**